pkg/issuer/acme/dns/execute/execute.go

```diff
@@ -0,0 +1,143 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package execute
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"path"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/jetstack/cert-manager/pkg/issuer/acme/dns/util"
+)
+
+type DNSProvider struct {
+	pluginPath       string
+	config           []byte
+	dns01Nameservers []string
+}
+
+func NewDNSProviderCredentials(pluginName string, pluginDirectory string, configJson []byte, nameservers []string) (*DNSProvider, error) {
+	if pluginName == "" {
+		return nil, fmt.Errorf("execute plugin name missing")
+	} else if pluginDirectory == "" {
+		return nil, fmt.Errorf("execute plugin directory name missing")
+	}
+
+	return &DNSProvider{
+		pluginPath:       path.Join(pluginDirectory, pluginName),
+		config:           configJson,
+		dns01Nameservers: nameservers,
+	}, nil
+}
+
+// Timeout returns the timeout and interval to use when checking for DNS
+// propagation. Adjusting here to cope with spikes in propagation times.
+func (c *DNSProvider) Timeout() (timeout, interval time.Duration) {
+	return 120 * time.Second, 2 * time.Second
+}
+
+func (c *DNSProvider) Present(domain, token, keyAuth string) error {
+	fqdn, value, ttl, err := util.DNS01Record(domain, keyAuth, c.dns01Nameservers)
+	if err != nil {
+		return err
+	}
+
+	envs := map[string]string{"EXECUTE_PLUGIN_ACTION": "present",
+		"EXECUTE_PLUGIN_DOMAIN": util.UnFqdn(fqdn),
+		"EXECUTE_PLUGIN_VALUE":  value,
+		"EXECUTE_PLUGIN_TTL":    string(ttl)}
+	err, _, _ = execute(c.pluginPath, envs, c.config, 0)
+
+	return err
+}
+
+func (c *DNSProvider) CleanUp(domain, token, keyAuth string) error {
+	fqdn, value, ttl, err := util.DNS01Record(domain, keyAuth, c.dns01Nameservers)
+	if err != nil {
+		return err
+	}
+
+	envs := map[string]string{"EXECUTE_PLUGIN_ACTION": "cleanup",
+		"EXECUTE_PLUGIN_DOMAIN": util.UnFqdn(fqdn),
+		"EXECUTE_PLUGIN_VALUE":  value,
+		"EXECUTE_PLUGIN_TTL":    strconv.Itoa(ttl)}
+	err, _, _ = execute(c.pluginPath, envs, c.config, 0)
+
+	return err
+}
+
+func execute(binaryPath string, envs map[string]string, config []byte, timeoutSeconds time.Duration) (error, string, string) {
+	var timeout time.Duration
+	if timeoutSeconds == 0 {
+		timeout = 30 * time.Second
+	} else {
+		timeout = timeoutSeconds * time.Second
+	}
+
+	var stdoutBuf, stderrBuf bytes.Buffer
+	cmd := exec.Cmd{
+		Path:        binaryPath,
+		Env:         envMapToSlice(envs),
+		Dir:         os.TempDir(),
+		Stdin:       bytes.NewReader(config),
```

Comment:
Instead of passing secret data directly to the plugin, we should define our own structure and encode that to JSON here.

Something like:

```
type ExecuteContext struct {
  // Identifier is the record that should have a TXT record set for, e.g. _acme-challenge.example.com
  Identifier string `json:"identifier"`

  // Key is the value that the TXT record should hold for this domain
  Key string `json:"key"`

  // Metadata is arbitrary additional metadata passed to the plugin from the Issuer resource.
  // This may contain a reference to a secret resource, containing secret data specific to this
  // configuration of the plugin.
  Metadata map[string]string `json:"metadata"`
}
```

This is only written off the top of my head, and may be missing some important fields...!

The output format should *also* be defined as part of this package too, and we should expect plugins to respond with this output format (even if that format is trivial for now).

This is basically then defining an API that we use to talk with external DNS providers, doing it this way should mean we can extend that API in future 😄